reader.go

```diff
@@ -705,15 +710,116 @@ func (d *decoder) parseDataFields(dm *defmsg, knownMsg bool, msgv reflect.Value)
 		if err != nil {
 			return reflect.Value{}, fmt.Errorf("error parsing data developer message: %v (field %d [%v] for [%v])", err, i, ddfd, dm)
 		}
+		fieldDescMap, ok := d.fieldDescMsgs[ddfd.devDataIndex]
+		if !ok {
+			if d.debug {
+				d.opts.logger.Printf("could not find dev field description for dev data index %d dev field #%d", ddfd.devDataIndex, ddfd.fieldNum)
+			}
+			continue
+		}
+		fieldDesc, ok := fieldDescMap[ddfd.fieldNum]
+		if !ok {
+			if d.debug {
+				d.opts.logger.Printf("could not find dev field description for dev field #%d, dm=%v", ddfd.fieldNum, dm)
+			}
+			continue
+		}
+		value, err := d.parseDevField(dm, fieldDesc, ddfd)
+		if err != nil {
+			return reflect.Value{}, err
+		}
+		var fieldName string
+		if len(fieldDesc.FieldName) > 0 {
+			fieldName = fieldDesc.FieldName[0]
+		} else {
+			fieldName = ""
+		}
+		var units string
+		if len(fieldDesc.Units) > 0 {
+			units = fieldDesc.Units[0]
+		} else {
+			units = ""
+		}
+
+		devField := DeveloperField{
+			DeveloperDataIndex:    fieldDesc.DeveloperDataIndex,
+			FieldDefinitionNumber: fieldDesc.FieldDefinitionNumber,
+			BaseTypeId:            uint8(fieldDesc.FitBaseTypeId),
+			FieldName:             fieldName,
+			Units:                 units,
+			Value:                 value,
+		}
+		devFieldsMap := msgv.FieldByName("DeveloperFields")
+		devFieldsMap.SetMapIndex(reflect.ValueOf(fieldName), reflect.ValueOf(devField))
 	}
 
 	if knownMsg && !msgv.IsValid() {
 		panic("internal decoder error: parse data fields: known message, but not (reflect) valid")
 	}
 
+	if msgv.IsValid() && msgv.Type() == reflect.TypeOf(FieldDescriptionMsg{}) {
+		fieldMsg := msgv.Interface().(FieldDescriptionMsg)
+		if d.fieldDescMsgs == nil {
+			d.fieldDescMsgs = map[byte]map[byte]FieldDescriptionMsg{}
+		}
```

Comment:
this is what I was messaging about earlier. Although it is technically not a valid fit file, there is a chance this map isn't initialized the first time you read it above, since it gets initialized on the first field description message you come across. this map should be initialized in the `decode` private function where `unknownFields` map is also initialized.
also if you could combine the u8s into a u16 you could skip the double map entirely, since it looks like you are ignoring the developer message entirely

Reply:
👍 Moving the initialization makes sense. 

I prefer a double map to doing bitwise operations which are IMO not readable.